torch/csrc/autograd/python_saved_variable_hooks.cpp

```diff
@@ -46,16 +46,22 @@ namespace torch { namespace autograd {
   }
 
   // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-  PyObject* PyDefaultSavedVariableHooks::pack_hook_ = nullptr;
+  std::atomic<PyObject*> PyDefaultSavedVariableHooks::pack_hook_(nullptr);
   // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-  PyObject* PyDefaultSavedVariableHooks::unpack_hook_ = nullptr;
+  std::atomic<PyObject*> PyDefaultSavedVariableHooks::unpack_hook_(nullptr);
 
   void PyDefaultSavedVariableHooks::set_hooks(py::function &pack_hook, py::function &unpack_hook) {
+    PyObject* expected = nullptr;
     TORCH_CHECK(!pack_hook_ && !unpack_hook_,
         "Setting default hooks but they have already been set. "
         "Hint: only one pair of hooks is allowed at a time.");
-    pack_hook_ = pack_hook.release().ptr();
-    unpack_hook_ = unpack_hook.release().ptr();
+    PyObject* pack_hook_ptr = pack_hook.release().ptr();
+    PyObject* unpack_hook_ptr = unpack_hook.release().ptr();
+    if (!pack_hook_.compare_exchange_strong(expected, pack_hook_ptr) ||
```

Comment:
What happens if one succeed and one fails here?
Both threads would end up in the if statement and raise an error. And so this is ok?

Reply:
I expect that if any of the exchange fails, then the program will crash, raising an error. 
If the program is run again, most likely the TORCH_CHECK from line 55 will fail and indicate the user that something is wrong with his code (e.g. he is setting default hooks in the forward pass of a module that is passed to `torch.nn.DataParallel`).
We want to fail fast here and I expect this race condition to be so rare that if it does happen in practice, it's worth opening an issue?

Reply:
The mutex-based solution seems easier to reason about; with the mutex we could convince ourselves that if two threads are trying to set the global hook, exactly one of them will fail.

I'm having a hard time reasoning if there are problems with reference counts if two threads end up inside this conditional and call reset_hooks(). A user's program doesn't have to terminate if one of the threads raise an exception, right?

Reply:
Yes, my mistake. Now I've convinced myself that the current code is problematic. The mutex-based solution looks appealing.